### ISSUE TITLE
Generated Gradle API JAR is different and causes cache misses

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#16790]
There are actually two problems here:

1. It seems that the Gradle API JAR in some cases is generated slightly differently. This shouldn't happen as we take pains to enforce the same file dates, file order etc.
2. `JavaCompile` tasks consuming this API JAR produce cache misses, even though the differences in the JAR itself only affect metadata, and thus should be normalized away by `@CompileClasspath` normalization.

We should figure out the cause of both problems. It is possible that we'll end up splitting this issue in two once we better understand what's happening.

---
cc: @gradle/execution

**Comment by lptr** [comment]
We checked the MD5 of `.gradle/caches/7.0-rc-1/generated-gradle-jars/gradle-api-7.0-rc-1.jar` on a few computers:

- for @melix's Linux it was `3c81e717e88af360a44646a34d7eb602`
- for @lptr's macOS it was also `3c81e717e88af360a44646a34d7eb602`
- on `dev78.gradle.org` and `dev118.gradle.org` (both Linux CI machines) it was `1f511bb`

**Comment by lptr** [comment]
The situation is similar with the most recent nightly being used (`7.1-20210328220041+0000`)

- all Linux CI agents have `b5ac1b8794104854d498a27875de9cfc`
- all Windows CI agents have `32ec7ce90e407ffccda8b162036234a8`
- all macOS CI agents have `32ec7ce90e407ffccda8b162036234a8`, too
- @lptr's macOS also has `32ec7ce90e407ffccda8b162036234a8`

So it looks like our Linux CI agents have something special going on, but not because they are Linux (as @melix's machine is also producing the non-special result).

**Comment by lptr** [comment]
We extracted the generated JARs on all platforms, and the contents of the extracted files are byte-for-byte the same.

We also checked with `unzip -t` and the metadata and file order is exactly the same between the different variants of the JAR.

**Comment by lptr** [comment]
We have seen that the difference between two variants are in the "extra info" part of the ZIP entries:

```text
--- local-data.txt	2021-03-23 09:17:33.000000000 +0100
+++ remote-data.txt	2021-03-23 09:17:40.000000000 +0100
@@ -101,7 +101,7 @@
 05542640  72 67 2f 67 72 61 64 6c  65 2f 63 6f 6e 66 69 67  |rg/gradle/config|
 05542650  75 72 61 74 69 6f 6e 63  61 63 68 65 2f 73 65 72  |urationcache/ser|
 05542660  69 61 6c 69 7a 61 74 69  6f 6e 2f 43 6f 6d 62 69  |ialization/Combi|
-05542670  6e 61 74 6f 72 73 4b 74  2e 63 6c 61 73 73 01 f3  |natorsKt.class..|
+05542670  6e 61 74 6f 72 73 4b 74  2e 63 6c 61 73 73 00 f3  |natorsKt.class..|
 05542680  7f 0c 80 ca fe ba be 00  00 00 34 02 de 01 00 39  |..........4....9|
 05542690  6f 72 67 2f 67 72 61 64  6c 65 2f 63 6f 6e 66 69  |org/gradle/confi|
 055426a0  67 75 72 61 74 69 6f 6e  63 61 63 68 65 2f 73 65  |gurationcache/se|
```

The difference is in the byte that comes right after the file name being `00` vs `01` here.

The next difference is that in the CI variant there’s an extra `01 00  00 ff ff` sequence added at `0554a676`.

**Comment by wolfs** [comment]
It seems like only `CompileClasspath` normalization is affected, given that no `compileKotlin` tasks show up with differences in this build comparison: https://ge.gradle.org/c/l5xewdvihpplm/kojhwsxervghs/task-inputs?expanded=WyJlb2E0dDR3Mnp5dHdlLWNsYXNzcGF0aCIsImVhZ2d5eDM2dWg3bDYtY2xhc3NmaWxlcyRrb3RsaW5fZ3JhZGxlX3BsdWdpbiIsImVhZ2d5eDM2dWg3bDYtc291cmNlc2V0b3V0cHV0Y2xhc3Nlc2RpciRrb3RsaW5fZ3JhZGxlX3BsdWdpbiJd&task-text=compile

**Comment by melix** [comment]
Some more insights. I have created [a branch](https://github.com/gradle/gradle/tree/cc/gradleapi-logging) which logs the individual md5 hashs of both generated API classes and classpath snapshotting.

I have run this branch locally using a dummy project containing:

.settings.gradle.kts
```kotlin
rootProject.name = "toto"
```

.build.gradle
```groovy
println "hello"
```

I executed this locally, which gave me this API jar: 

```
$ ls -al homedir/caches/7.1-20210413075535+0000/generated-gradle-jars/gradle-api-7.1-20210413075535+0000.jar
-rw-rw-r-- 1 cchampeau cchampeau 148603383 Apr 13 09:56 homedir/caches/7.1-20210413075535+0000/generated-gradle-jars/gradle-api-7.1-20210413075535+0000.jar
```

then I did the same on dev69:

```
$ ls -al gradlehome/caches/7.1-20210413075535+0000/generated-gradle-jars/
-rw-rw-r-- 1 cchampeau cchampeau 148603388 Apr 13 13:59 gradle-api-7.1-20210413075535+0000.jar
```

So the good news is that the jars are different, so the behavior is _reproducible_. However, I compared the log files, and they are _identical_. This means that whatever is generated by _classpath_ snapshotting or ASM manipulation is _identical_.

This leaves us with resources, which aren't logged in my branch.

**Comment by melix** [comment]
So I have found the reason why the generated jars are different on CI and locally. I was looking at the source code of Java's `java.util.zip.Deflater` class and noticed that there were native calls. The comments said that there was the native `zlib` library involved.

So I looked at what is the version of zlib on CI: 

```
$ apt show zlib1g
Package: zlib1g
Version: 1:1.2.8.dfsg-2ubuntu4.3
...
```

then I checked what I had locally:

```
Package: zlib1g
Version: 1:1.2.11.dfsg-2ubuntu1.2
```

Next step was to create a docker container using Ubuntu 16.04 LTS. I installed JDK 11 on it, then Gradle 7, generated the API jars, which got me with a jar with this hash: 

```
$ gradle -g ubuntu16-base help --no-daemon
...
$ md5sum ubuntu16-base/caches/7.0/generated-gradle-jars/gradle-api-7.0.jar 
2b33ffe5cafa32e54b411a636e952516  ubuntu16-base/caches/7.0/generated-gradle-jars/gradle-api-7.0.jar
```

Then I downloaded a more recent version of zlib and installed it in my Docker image:

```
dpkg -i zlib1g_1.2.11.dfsg-2ubuntu1.2_amd64.deb
```

and executed the same build again, using a different home:
```
$ gradle -g ubuntu16-zlib-upgradle help --no-daemon
...
$ md5sum ubuntu16-zlib-upgradle/caches/7.0/generated-gradle-jars/gradle-api-7.0.jar 
f58bcae3c212bd47161c97b2b330c497  ubuntu16-zlib-upgradle/caches/7.0/generated-gradle-jars/gradle-api-7.0.jar
```

Ah ah! This time the hash is different! Last step was to compare with the hash I had locally with Gradle 7, and it turned out to be `f58bcae3c212bd47161c97b2b330c497`.

**Conclusion**: we generate different jars because Java delegates to zlib at some point, and this library generates different archives for different versions.

This means, technically, that without normalization, it is very well possible to get cache misses because of "byte-to-byte, expanded archives" but actually in a different form.
